File: app/schemas/complex_query.py
from typing import List, Optional
from pydantic import BaseModel


class ComplexQueryCondition(BaseModel):
    """Условие внутри сложного запроса"""
    field: str  # title, content, author, source, publication_title, url, doi, tag
    operator: str = "contains"  # contains, equals, starts_with
    value: str
    negated: bool = False  # НЕ


class ComplexQueryGroup(BaseModel):
    """Группа условий с оператором"""
    operator: str = "AND"  # AND, OR
    conditions: List[ComplexQueryCondition] = []
    groups: List["ComplexQueryGroup"] = []


# Обновляем рекурсивную ссылку
ComplexQueryGroup.model_rebuild()


class ComplexQuery(BaseModel):
    """Корень сложного запроса"""
    root: ComplexQueryGroup

File: frontend/src/types/complexQuery.ts
export type ConditionField =
  | 'title'
  | 'content'
  | 'author'
  | 'source'
  | 'publication_title'
  | 'url'
  | 'doi'
  | 'tag'

export type ConditionOperator = 'contains' | 'equals' | 'starts_with'

export interface ComplexQueryCondition {
  field: ConditionField
  operator: ConditionOperator
  value: string
  negated: boolean
}

export interface ComplexQueryGroup {
  operator: 'AND' | 'OR'
  conditions: ComplexQueryCondition[]
  groups: ComplexQueryGroup[]
}

export interface ComplexQuery {
  root: ComplexQueryGroup
}

export function createEmptyGroup(operator: 'AND' | 'OR' = 'AND'): ComplexQueryGroup {
  return {
    operator,
    conditions: [],
    groups: [],
  }
}

export function createEmptyCondition(): ComplexQueryCondition {
  return {
    field: 'title',
    operator: 'contains',
    value: '',
    negated: false,
  }
}